netbox/page.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2017 The go-netbox Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package netbox
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+// A Page contains all necessary information about a the current position
+// in a paged result. It is used to walk over all pages from List calls.
+type Page struct {
+	limit    int
+	offset   int
+	done     bool
+	c        *Client
+	data     pageData
+	endpoint string
+	options  Valuer
+	err      error
+}
+
+// NewPage Returns a new Page to walk over List calls
+func NewPage(client *Client, endpoint string, options Valuer) *Page {
+	return &Page{limit: 50, // netbox server default
+		offset:   0,
+		c:        client,
+		done:     false,
+		data:     pageData{},
+		endpoint: endpoint,
+		options:  options,
+		err:      nil}
+}
+
+// Internal representation of a page
+type pageData struct {
+	Count       int             `json:"count"`
+	NextURL     string          `json:"next"`
+	PreviousURL string          `json:"previous"`
+	Results     json.RawMessage `json:"results"`
+}
+
+// values implements the Valuer interface. One could pass options
+// to a List call. These must be extended by limit and offset to
+// get the appropriate next page.
+func (p *Page) values() (url.Values, error) {
+	if p == nil {
+		return nil, errors.New("page not defined")
+	}
+	var v url.Values
+	if p.options == nil {
+		v = url.Values{}
+	} else {
+		var err error
+		v, err = p.options.values()
+		if err != nil {
+			return nil, err
+		}
+		if v == nil {
+			v = url.Values{}
+		}
+	}
+	v.Set("limit", strconv.Itoa(p.limit))
+	v.Set("offset", strconv.Itoa(p.offset))
+
+	return v, nil
+}
+
+// Next is what you call in a for loop, to iterate over all pages
+// for myPage.Next() { ... }
+// Next performs the real query and sets the next page
+func (p *Page) Next() bool {
+	if p == nil {
+		return false
+	}
+	if p.done {
+		return false
+	}
+
+	req, err := p.c.NewRequest(http.MethodGet, p.endpoint, p)
+	if err != nil {
+		p.err = err
+		return false
+	}
+
+	// Clear NextURL, because it will not be overriden if it is empty in the
+	// resultset. I(cglaubitz) do not want to allocate a new PageData on each call
+	// to Next() to reduce the need for garbage collection. No idea if this still
+	// works when I have to "empty" NextURL here.
```

Comment:
I wouldn't worry too much about making more garbage. IMO optimize for simplicity and readability first, but if you have a memory profile that shows that this produces too much garbage, it can be reworked later.